### PR TITLE
AmazonSearch Scenario

### PR DESCRIPTION
**Pull request by MedhaGupta-Ril**
Adding scenario related to amazon search using teswiz framework

### CODE REVIEW

File: src/test/java/com/znsio/sample/e2e/screen/web/amazon/AmazonProductViewWeb.java [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+        driver.waitForClickabilityOf(firstProduct).click();
+
+        return AmazonProductViewScreen.get();
+    }
+
+    @Override
+    public AmazonProductViewScreen addProductToCart(){
+        driver.switchToNextTab();
+        driver.waitForClickabilityOf(addCartButton).click();
+        waitFor(5);
```
Comment (mohammadsrk) [inline src/test/java/com/znsio/sample/e2e/screen/web/amazon/AmazonProductViewWeb.java:46]: Why hard wait here? Avoid using hard waits

File: src/test/java/com/znsio/sample/e2e/screen/web/amazon/AmazonProductViewWeb.java [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+        return closeSideBar();
+    }
+
+    @Override
+    public AmazonProductViewScreen closeSideBar(){
+        driver.waitForClickabilityOf(sideBarCloseButton).click();
+        return AmazonProductViewScreen.get();
+    }
```
Comment (mohammadsrk) [inline src/test/java/com/znsio/sample/e2e/screen/web/amazon/AmazonProductViewWeb.java:54]: Use return this;

File: src/test/java/com/znsio/sample/e2e/screen/web/amazon/AmazonProductViewWeb.java [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+
+        return AmazonProductViewScreen.get();
+    }
+
+    @Override
+    public AmazonProductViewScreen addProductToCart(){
+        driver.switchToNextTab();
+        driver.waitForClickabilityOf(addCartButton).click();
+        waitFor(5);
+        return closeSideBar();
```
Comment (mohammadsrk) [inline src/test/java/com/znsio/sample/e2e/screen/web/amazon/AmazonProductViewWeb.java:47]: call closeSideBar inside the method and return the current object to increase code readability.

File: src/test/java/com/znsio/sample/e2e/screen/web/amazon/AmazonProductViewWeb.java [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+import org.openqa.selenium.WebElement;
+
+import static com.znsio.e2e.tools.Wait.waitFor;
+
+public class AmazonProductViewWeb extends AmazonProductViewScreen {
+    private final Driver driver;
+    private final Visual visually;
+    private static final String SCREEN_NAME = AmazonHomeScreenWeb.class.getSimpleName();
+    private static final Logger LOGGER = Logger.getLogger(SCREEN_NAME);
+    private static final String NOT_YET_IMPLEMENTED = " not yet implemented";
```
Comment (mohammadsrk) [inline src/test/java/com/znsio/sample/e2e/screen/web/amazon/AmazonProductViewWeb.java:17]: remove unwanted , unused line of code.

File: src/test/java/com/znsio/sample/e2e/screen/web/amazon/AmazonHomeScreenWeb.java [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+    private static final By bySearchBoxXpath = By.id("twotabsearchtextbox");
+    private static final By bySearchIconClassName = By.cssSelector("[value='Go']");
+
+    public AmazonHomeScreenWeb(Driver driver, Visual visually) {
+        this.driver = driver;
+        this.visually = visually;
+    }
+
+    @Override
+    public AmazonProductViewScreen searchProduct(String productName){
```
Comment (mohammadsrk) [inline src/test/java/com/znsio/sample/e2e/screen/web/amazon/AmazonHomeScreenWeb.java:29]: visuall check is missing in all screens.

File: src/test/java/com/znsio/sample/e2e/businessLayer/Amazon/AmazonShoppingCartBL.java [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+        long threadId = Thread.currentThread()
+                .getId();
+        this.context = Runner.getTestExecutionContext(threadId);
+        softly = Runner.getSoftAssertion(threadId);
+        this.currentUserPersona = SAMPLE_TEST_CONTEXT.GUEST_USER;
+        this.currentPlatform = Runner.platform;
+        LOGGER.info("AmazonSearchBL created");
+    }
+
+    public AmazonShoppingCartBL openShoppingCart(){
```
Comment (mohammadsrk) [inline src/test/java/com/znsio/sample/e2e/businessLayer/Amazon/AmazonShoppingCartBL.java:31]: improve method name

File: src/test/java/com/znsio/sample/e2e/businessLayer/Amazon/AmazonShoppingCartBL.java [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+        LOGGER.info(String.format("User go into the shopping cart"));
+        AmazonShoppingCartScreen.get().openShoppingCart();
+        return this;
+    }
+    public AmazonShoppingCartBL productIsVisibleInCart(String productName){
```
Comment (mohammadsrk) [inline src/test/java/com/znsio/sample/e2e/businessLayer/Amazon/AmazonShoppingCartBL.java:36]: improve method name

File: src/test/java/com/znsio/sample/e2e/businessLayer/Amazon/AmazonProductViewBL.java [hunk 8]
```diff
@@ -0,0 +1,40 @@
...
+        long threadId = Thread.currentThread()
+                .getId();
+        this.context = Runner.getTestExecutionContext(threadId);
+        softly = Runner.getSoftAssertion(threadId);
+        this.currentUserPersona = SAMPLE_TEST_CONTEXT.GUEST_USER;
+        this.currentPlatform = Runner.platform;
+        LOGGER.info("AmazonSearchBL created");
+    }
+
+    public AmazonProductViewBL selectFirstProduct(){
```
Comment (mohammadsrk) [inline src/test/java/com/znsio/sample/e2e/businessLayer/Amazon/AmazonProductViewBL.java:29]: incorrect BL method name

File: src/test/java/com/znsio/sample/e2e/screen/android/amazon/AmazonHomeScreenAndroid.java [hunk 9]
```diff
@@ -0,0 +1,64 @@
...
+public class AmazonHomeScreenAndroid extends AmazonHomeScreen {
+
+    private final Driver driver;
+    private final Visual visually;
+    private static final String SCREEN_NAME = CalculatorScreenAndroid.class.getSimpleName();
+    private static final Logger LOGGER = Logger.getLogger(SCREEN_NAME);
+
+    private static By searchArea = By.xpath("//android.widget.EditText");
+    private static String searchButtonXpath = "//android.view.View[@resource-id = \"nav-search-form\"]//android.widget.Button";
+
```
Comment (mohammadsrk) [inline src/test/java/com/znsio/sample/e2e/screen/android/amazon/AmazonHomeScreenAndroid.java:32]: xpath variables naming convention is not correct.

File: src/test/java/com/znsio/sample/e2e/screen/android/amazon/AmazonHomeScreenAndroid.java [hunk 10]
```diff
@@ -0,0 +1,64 @@
...
+        visually.checkWindow(SCREEN_NAME, "Product list view");
+        String productName = driver.waitTillElementIsPresent(firstProduct).getText();
+        return productName;
+    }
+
+    @Override
+    public int getproductCount(){
+        int count = driver.findElements(countElement).size();
+        return count;
+    }
```
Comment (mohammadsrk) [inline src/test/java/com/znsio/sample/e2e/screen/android/amazon/AmazonHomeScreenAndroid.java:63]: Loggers are missing.

File: src/test/java/com/znsio/sample/e2e/screen/android/amazon/AmazonProductViewAndroid.java [hunk 11]
```diff
@@ -0,0 +1,66 @@
...
+    private final Driver driver;
+    private final Visual visually;
+    private static final String SCREEN_NAME = AmazonHomeScreenWeb.class.getSimpleName();
+    private static final Logger LOGGER = Logger.getLogger(SCREEN_NAME);
+
+    private static By firstProduct = By.xpath("//android.view.View[contains(@content-desc, \""+ SAMPLE_TEST_CONTEXT.PRODUCT_NAME+"\")]/android.view.View[2]");
+    private static final By productTitle = By.xpath("//android.view.View[@resource-id = \"title\"]");
+    private static By addCartButton = By.xpath("//android.widget.Button[@resource-id = \"add-to-cart-button\"]");
+    private static final By addedToCartMsg = By.xpath("//div[@id='attachDisplayAddBaseAlert']/span");
+
```
Comment (mohammadsrk) [inline src/test/java/com/znsio/sample/e2e/screen/android/amazon/AmazonProductViewAndroid.java:26]: xpath variables naming convention is not correct.

File: src/test/java/com/znsio/sample/e2e/screen/android/amazon/AmazonProductViewAndroid.java [hunk 12]
```diff
@@ -0,0 +1,66 @@
...
+    public boolean verifyProductDetails(){
+        WebElement element = driver.waitTillElementIsPresent(productTitle);
+        return element.isDisplayed();
+    }
+
+    @Override
+    public AmazonProductViewScreen clickAddToCartButton(){
+        LOGGER.info(String.format("Add product to the cart"));
+        while(true) {
+            waitFor(1);
```
Comment (mohammadsrk) [inline src/test/java/com/znsio/sample/e2e/screen/android/amazon/AmazonProductViewAndroid.java:50]: No Hard waits allowed.

File: src/test/java/com/znsio/sample/e2e/screen/android/amazon/AmazonProductViewAndroid.java [hunk 13]
```diff
@@ -0,0 +1,66 @@
...
+            if(driver.isElementPresent(addCartButton))
+                break;
+            driver.scrollDownByScreenSize();
```
Comment (mohammadsrk) [inline src/test/java/com/znsio/sample/e2e/screen/android/amazon/AmazonProductViewAndroid.java:53]: What happens if it reaches the end of the screen, will loop ever ends?

File: src/test/java/com/znsio/sample/e2e/screen/android/amazon/AmazonShoppingCartAndroid.java [hunk 14]
```diff
@@ -0,0 +1,47 @@
...
+    private final Driver driver;
+    private final Visual visually;
+    private static final String SCREEN_NAME = AmazonHomeScreenWeb.class.getSimpleName();
+    private static final Logger LOGGER = Logger.getLogger(SCREEN_NAME);
+    private static final String NOT_YET_IMPLEMENTED = " not yet implemented";
+    private static final By shoppingCartButton = By.xpath("//android.view.View[@content-desc=\"Cart\"]/android.view.View");
+
+    private static final By shoppingCartMessage = By.xpath("//android.view.View[@resource-id = \"sc-buy-box\"]");
+    private static final By firstProductInCart = By.xpath("//android.view.View[contains(@content-desc=\""+ SAMPLE_TEST_CONTEXT.PRODUCT_NAME+"\")]/android.widget.TextView");
+
```
Comment (mohammadsrk) [inline src/test/java/com/znsio/sample/e2e/screen/android/amazon/AmazonShoppingCartAndroid.java:23]: xpath variables naming convention is not correct.

File: src/test/java/com/znsio/sample/e2e/screen/android/amazon/AmazonHomeScreenAndroid.java (hunk 9 above)
Comment (MedhaGupta-Ril) [reply src/test/java/com/znsio/sample/e2e/screen/android/amazon/AmazonHomeScreenAndroid.java:32]: corrected

File: src/test/java/com/znsio/sample/e2e/screen/android/amazon/AmazonProductViewAndroid.java (hunk 11 above)
Comment (MedhaGupta-Ril) [reply src/test/java/com/znsio/sample/e2e/screen/android/amazon/AmazonProductViewAndroid.java:26]: corrected

File: src/test/java/com/znsio/sample/e2e/screen/android/amazon/AmazonShoppingCartAndroid.java (hunk 14 above)
Comment (MedhaGupta-Ril) [reply src/test/java/com/znsio/sample/e2e/screen/android/amazon/AmazonShoppingCartAndroid.java:23]: corrected

File: src/test/java/com/znsio/sample/e2e/screen/android/amazon/AmazonHomeScreenAndroid.java (hunk 10 above)
Comment (MedhaGupta-Ril) [reply src/test/java/com/znsio/sample/e2e/screen/android/amazon/AmazonHomeScreenAndroid.java:63]: Resolved

File: src/test/java/com/znsio/sample/e2e/screen/android/amazon/AmazonProductViewAndroid.java (hunk 12 above)
Comment (MedhaGupta-Ril) [reply src/test/java/com/znsio/sample/e2e/screen/android/amazon/AmazonProductViewAndroid.java:50]: Resolved

File: src/test/java/com/znsio/sample/e2e/screen/android/amazon/AmazonProductViewAndroid.java (hunk 13 above)
Comment (MedhaGupta-Ril) [reply src/test/java/com/znsio/sample/e2e/screen/android/amazon/AmazonProductViewAndroid.java:53]: Resolved

File: caps/amazonsearch_local_capabilities.json [hunk 15]
```diff
@@ -0,0 +1,29 @@
+{
+  "android": {
+    "adbExecTimeout": 45000,
+    "app": {
+      "local": "/Users/medha.gupta/Desktop/NewFolder/getting-started-with-teswiz/temp/Amazon.apk"
```
Comment (vishal07111998) [inline caps/amazonsearch_local_capabilities.json:5]: path should be generic

File: src/test/java/com/znsio/sample/e2e/businessLayer/Amazon/AmazonHomeBL.java [hunk 16]
```diff
@@ -0,0 +1,41 @@
...
+import com.znsio.sample.e2e.screen.amazonsearch.AmazonHomeScreen;
+import org.apache.log4j.Logger;
+import org.assertj.core.api.SoftAssertions;
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+public class AmazonHomeBL {
+    private static final Logger LOGGER = Logger.getLogger(AmazonHomeBL.class.getName());
+    private final TestExecutionContext context;
+    private final SoftAssertions softly;
```
Comment (vishal07111998) [inline src/test/java/com/znsio/sample/e2e/businessLayer/Amazon/AmazonHomeBL.java:16]: remove unused variable

File: src/test/java/com/znsio/sample/e2e/businessLayer/Amazon/AmazonProductViewBL.java [hunk 17]
```diff
@@ -0,0 +1,55 @@
...
+        this.context = Runner.getTestExecutionContext(threadId);
+        softly = Runner.getSoftAssertion(threadId);
+        this.currentUserPersona = SAMPLE_TEST_CONTEXT.GUEST_USER;
+        this.currentPlatform = Runner.platform;
+        LOGGER.info("AmazonSearchBL created");
+    }
+
+    public AmazonProductViewBL selectProduct(){
+        LOGGER.info(String.format("Select the first product"));
+        Boolean correct_product = AmazonProductViewScreen.get().selectFirstProduct()
```
Comment (vishal07111998) [inline src/test/java/com/znsio/sample/e2e/businessLayer/Amazon/AmazonProductViewBL.java:33]: no need of variable correct_product
also follow the same naming convention

File: src/test/java/com/znsio/sample/e2e/businessLayer/Amazon/AmazonProductViewBL.java [hunk 18]
```diff
@@ -0,0 +1,55 @@
...
+    }
+    
+    public AmazonProductViewBL prepareShoppingCart(){
+        LOGGER.info(String.format("User add product into the shopping cart"));
+        String shoppingCartMessage = context.getTestStateAsString(SAMPLE_TEST_CONTEXT.ADDED_TO_CART_MESSAGE);
+        AmazonProductViewScreen.get().clickAddToCartButton();
+        String actualAddedToCartMessage;
+        switch (currentPlatform){
+            case android:
+//                actualAddedToCartMessage = AmazonProductViewScreen.get().getAddedToCartMessage();
```
Comment (vishal07111998) [inline src/test/java/com/znsio/sample/e2e/businessLayer/Amazon/AmazonProductViewBL.java:46]: remove commented code

File: src/test/java/com/znsio/sample/e2e/screen/android/amazon/AmazonProductViewAndroid.java [hunk 19]
```diff
@@ -0,0 +1,61 @@
...
+import org.openqa.selenium.JavascriptExecutor;
+import org.openqa.selenium.WebElement;
+
+import static com.znsio.e2e.tools.Wait.waitFor;
+
+public class AmazonProductViewAndroid extends AmazonProductViewScreen{
+
+    private final Driver driver;
+    private final Visual visually;
+    private static final String SCREEN_NAME = AmazonHomeScreenWeb.class.getSimpleName();
```
Comment (vishal07111998) [inline src/test/java/com/znsio/sample/e2e/screen/android/amazon/AmazonProductViewAndroid.java:19]: screen name is different

File: src/test/java/com/znsio/sample/e2e/businessLayer/Amazon/AmazonProductViewBL.java (hunk 17 above)
Comment (mshahrukh2) [inline src/test/java/com/znsio/sample/e2e/businessLayer/Amazon/AmazonProductViewBL.java:33]: @MedhaGupta-Ril Please incorporate the above comments, and in place of the Boolean class use its primitive type.

**Pull request closed without merging**